### PR TITLE
Minor improvements

### PR DESCRIPTION
**Pull request by anandabits**
You have done a nice job of refining the original draft so I don't have a lot to add.  I think the motivating example is stronger when `compactMap` is used in the example as that unwraps the associated value which is usually what you would want when filtering an array of results.  The rest of the edits are minor.

There is one more substantive issue to consider so I'm just providing feedback here (at least for now) rather than making edits directly.  I think there needs to be a way to opt-out of synthesis.  I can't think of a good reason why someone would want to do that but I do think programmers should be in control.  It would also be nice if the programmer could specify access control for the synthesized properties (they should not be `public` by default).

The easiest way to do that would be to require the user to opt-in to synthesis somehow (opting out by default).  I don't have a great idea for how to make that work though.  A protocol seems inappropriate as there are not requirements that are being fulfilled by the synthesized code.

The closest existing language feature I can think of is memberwise and default initializer synthesis for structs.  These are synthesized by default and disabled by the presence of an initializer in the initial declaration of the struct.  Something similar _could_ work here: the properties are synthesized unless the initial declaration of the enum contains a computed property.  However, I don't think that is appropriate.  There are good reasons for suppressing initializer synthesis in the presence of a manual initializer that don't apply to synthesized properties.

A related topic is access control for the synthesized properties.  Synthesized initializers are `internal`, following the principle that all public API should be explicitly stated by a library author.  The all-or-nothing nature of initializer synthesis is unfortunate and improving this is the goal I had with the flexible memberwise initializer proposal (which I hope we revisit someday).  Note: the access control issue isn't relevant for synthesis of protocol requirements as Swift only synthesizes conformances for `public` protocols and the visibility of a conformance must currently be at least as great as the protocol.

It would be nice if we can find an approach that allows programmers to opt-out of property synthesis if necessary and also opt-in to making the synthesized properties `public` where desired.  I don't have anything concrete in mind yet but will give it some thought over the next few days.  I'm interested in hearing your thoughts on these issues.

### CODE REVIEW

File: proposals/0000-automatically-derive-properties-for-enum-cases.md [hunk 1]
```diff
@@ -44,11 +44,11 @@ if case let .value(value) = result {
 This challenge increases with closures. Swift privileges single-line closures with both implicit returns and better-inferred return types. It should not take multiple lines to extract associated value information from enum instances. The current technology requires excessive boilerplate and ceremony to access these values.
 
 ``` swift
-array.filter { result -> Bool in
-    if case .value = result {
-        return true
+array.compactMap { result -> String in
```
Comment (stephencelis) [inline proposals/0000-automatically-derive-properties-for-enum-cases.md]: This should be `String?`, no?

File: proposals/0000-automatically-derive-properties-for-enum-cases.md [hunk 2]
```diff
@@ -74,10 +74,16 @@ Properties reduce code size and enhance readability. Contrast the following prop
...
 ``` swift
 // select items supporting associated value `value`
-array.filter { $0.value != nil }
+array.compactMap { $0.value }
 ```
 
-This approach is widely applicable. The Swift Standard Library provides many higher-order methods, including `map`, `filter`, `reduce`, `flatMap`, `sorted`, `first(where:)`, and more. Each is easier to use, read, and maintain when expressed as single lines.
+This approach is widely applicable. The Swift Standard Library provides many higher-order methods, including `map`, `filter`, `reduce`, `flatMap`, `compactMap`,  `sorted`, `first(where:)`, and more. Each is easier to use, read, and maintain when expressed as single lines.
+
+The benefit is arguably even more significant for APIs which provide KeyPath overloads (and more generally in the future should KeyPaths ever become a subtype of functions):
```
Comment (stephencelis) [inline proposals/0000-automatically-derive-properties-for-enum-cases.md:82]: Ha! It's almost as if you saw my other branch 😉

https://github.com/stephencelis/swift-evolution/blob/key-path-getter-promotion/proposals/0000-key-path-getter-promotion.md

**Comment by stephencelis** [comment]
Edits look good! Just a small catch on the first `compactMap` example. Also: be sure to add yourself to the authors if you'd like!

**Comment by stephencelis** [comment]
I don't know if I have a solution for opt-out (other than the fact that the compiler will have to skip generation for collisions, so someone could define `var myCaseName: Never { fatalError() }`, though this is a hack and obviously not a good solution). Let me think on it. If you come up with any pseudo-codey ideas, please lemme know!

As for access control level, it seems to make the most sense to me for the access control to be the same as the static cases, no? So `public enum` would make the properties `public` as well. I can't think of a use case for otherwise.

File: proposals/0000-automatically-derive-properties-for-enum-cases.md (hunk 1 above)
Comment (anandabits) [reply proposals/0000-automatically-derive-properties-for-enum-cases.md]: Oops!  😊

File: proposals/0000-automatically-derive-properties-for-enum-cases.md (hunk 2 above)
Comment (anandabits) [reply proposals/0000-automatically-derive-properties-for-enum-cases.md:82]: Awesome!  I didn't see that but this subtype relationship has been discussed in the past.

**Comment by anandabits** [comment]
@stephencelis good catch on the example!  I should know better than to write code in a Markdown editor! :)

I haven't thought of a great opt-out solution yet.  I'll keep thinking about it.

Regarding access control, the core team has been pretty reluctant to expose any `public` capability that wasn't explicitly requested by a library author.  I tend to agree with them on this.  

Here's a different way to approach this: an argument can be made that these properties should be a core capability of an enum case and should _always_ be available when the case is available.  Adopting that position would eliminate the opt-out and access control issues but would complicate migration a bit (users would need to remove manually written properties).  This position might be the best path to take as enum cases are already a declaration syntax that generates more than one usage capability (i.e. a factory method and a pattern).  What do you think about this direction?

**Comment by stephencelis** [comment]
I like (and probably prefer) it! Might make implementation more (or less) complicated. Unsure how diagnostics would factor in.

**Comment by anandabits** [comment]
This probably makes the implementation itself more straightforward - any conflicting manually declared properties become duplicate declaration warnings.  

Of course migration could become much more complex.  It should be possible to use AST patterns to detect manually written properties which have semantics identical to the synthesized properties and just remove them.  I'm not sure how far down that path implementers would want to go.  Fortunately that decision doesn't need to be made as part of the proposal process (although it would probably be part of the source compatibility consideration in the core team's decision).

I can't think of any good arguments _against_ this direction except for source compatibility.  Making these properties consistently available 100% of the time is a huge argument in its favor.  Unfortunately source compatibility is consideration these days.  Let's float this in the forums and see if anyone from the core team indicates whether it might be considered or not.  I'm about to head to the airport - feel free to post if you'd like before I have a chance.

**Pull request merged**